tls/extensions/s2n_client_server_name.c

```diff
@@ -22,67 +22,85 @@
-    GUARD(s2n_stuffer_write_uint16(out, server_name_len + 5));
+    return strlen(conn->server_name) > 0;
+}
 
-    /* Size of all of the server names */
-    GUARD(s2n_stuffer_write_uint16(out, server_name_len + 3));
+static int s2n_client_server_name_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    struct s2n_stuffer_reservation server_name_list_size;
+    GUARD(s2n_stuffer_reserve_uint16(out, &server_name_list_size));
 
-    /* Name type - host name, RFC3546 */
-    GUARD(s2n_stuffer_write_uint8(out, 0));
+    /* NameType, as described by RFC6066.
+     * host_name is currently the only possible NameType defined. */
+    GUARD(s2n_stuffer_write_uint8(out, S2N_NAME_TYPE_HOST_NAME));
 
-    struct s2n_blob server_name = {0};
-    server_name.data = (uint8_t *) conn->server_name;
-    server_name.size = server_name_len;
-    GUARD(s2n_stuffer_write_uint16(out, server_name_len));
-    GUARD(s2n_stuffer_write(out, &server_name));
+    GUARD(s2n_stuffer_write_uint16(out, strlen(conn->server_name)));
+    GUARD(s2n_stuffer_write_bytes(out, (const uint8_t *) conn->server_name, strlen(conn->server_name)));
 
-    return 0;
+    GUARD(s2n_stuffer_write_vector_size(server_name_list_size));
+    return S2N_SUCCESS;
 }
 
-int s2n_parse_client_hello_server_name(struct s2n_connection *conn, struct s2n_stuffer *extension)
+static int s2n_client_server_name_check(struct s2n_connection *conn, struct s2n_stuffer *extension, uint16_t *server_name_len)
 {
-    if (conn->server_name[0]) {
-        /* already parsed server name extension, exit early */
-        return 0;
-    }
-
     uint16_t size_of_all;
-    uint8_t server_name_type;
-    uint16_t server_name_len;
-    uint8_t *server_name;
-
     GUARD(s2n_stuffer_read_uint16(extension, &size_of_all));
-    if (size_of_all > s2n_stuffer_data_available(extension) || size_of_all < 3) {
-        /* the size of all server names is incorrect, ignore the extension */
-        return 0;
-    }
+    lte_check(size_of_all, s2n_stuffer_data_available(extension));
 
+    uint8_t server_name_type;
     GUARD(s2n_stuffer_read_uint8(extension, &server_name_type));
-    if (server_name_type != 0) {
-        /* unknown server name type, ignore the extension */
-        return 0;
-    }
+    eq_check(server_name_type, S2N_NAME_TYPE_HOST_NAME);
+
+    GUARD(s2n_stuffer_read_uint16(extension, server_name_len));
+    lt_check(*server_name_len, sizeof(conn->server_name));
+    lte_check(*server_name_len, s2n_stuffer_data_available(extension));
 
-    GUARD(s2n_stuffer_read_uint16(extension, &server_name_len));
-    if (server_name_len + 3 > size_of_all) {
-        /* the server name length is incorrect, ignore the extension */
-        return 0;
+    return S2N_SUCCESS;
+}
+
+static int s2n_client_server_name_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
+    /* Exit early if we've already parsed the server name */
+    if (conn->server_name[0]) {
```

Comment:
Why does this mean the server name is parsed?

Reply:
Because otherwise server_name[0] is 0, because the array is 0-initialized along with the rest of conn. If the name is parsed, server_name[0] won't be 0, because 0 is the null-terminator.

(I did not change this logic in the refactor)